x-pack/plugins/security_solution/public/resolver/models/resolver_tree.ts

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import {
+  ResolverTree,
+  ResolverEvent,
+  ResolverNodeStats,
+  ResolverLifecycleNode,
+} from '../../../common/endpoint/types';
+import { uniquePidForProcess } from './process_event';
+
+/**
+ * ResolverTree is a type returned by the server.
+ */
+
+/**
+ * This returns the 'LifecycleNodes' of the tree. These nodes have
+ * the entityID and stats for a process. Used by `relatedEventsStats`.
+ */
+function lifecycleNodes(tree: ResolverTree): ResolverLifecycleNode[] {
+  return [tree, ...tree.children.childNodes, ...tree.ancestry.ancestors];
+}
+
+/**
+ * All the process events
+ */
+export function lifecycleEvents(tree: ResolverTree) {
+  const events: ResolverEvent[] = [...tree.lifecycle];
+  for (const { lifecycle } of tree.children.childNodes) {
+    events.push(...lifecycle);
+  }
+  for (const { lifecycle } of tree.ancestry.ancestors) {
+    events.push(...lifecycle);
+  }
+  return events;
+}
+
+/**
+ * This returns a map of entity_ids to stats for the related events and alerts.
+ */
+export function relatedEventsStats(tree: ResolverTree): Map<string, ResolverNodeStats> {
+  const nodeStats: Map<string, ResolverNodeStats> = new Map();
+  for (const node of lifecycleNodes(tree)) {
+    if (node.stats) {
+      nodeStats.set(node.entityID, node.stats);
+    }
+  }
+  return nodeStats;
+}
+
+/**
+ * ResolverTree type is returned by the server. It organizes events into a complex structure. The
+ * organization of events in the tree is done to associate metadata with the events. The client does not
+ * use this metadata. Intsead, the client flattens the tree into an array. Therefore we can safely
+ * make a malformed RelsoverTree for the purposes of the tests, so long as it is flattened in a predicatable way.
```

Comment:
🤷 Resolver spelled wrong in the comment

Reply:
Intsead and predicatable too. (i turned on spellcheck lol)